packages/processing/pitop/processing/algorithms/ball_detect.py

```diff
@@ -162,181 +247,92 @@ def __init__(self, image_processing_width: int = 320, format: str = "OpenCV"):
-
-    def __call__(self, frame, color: Union[str, list] = "red"):
-        def parse_colors(color_arg):
-            colors = []
-            if isinstance(color_arg, str):
-                if color_arg not in VALID_COLORS:
-                    raise ValueError(
-                        f"Valid color values are {', '.join(VALID_COLORS[:-1])} or {VALID_COLORS[-1]}"
-                    )
-                colors = [color_arg]
-            elif type(color_arg) in (list, tuple):
-                if not set(color_arg).issubset(VALID_COLORS):
-                    raise ValueError(
-                        f"Valid color values are {', '.join(VALID_COLORS[:-1])} or {VALID_COLORS[-1]}"
-                    )
-                colors = color_arg
-            if len(colors) > 3:
-                raise ValueError("Cannot pass more than three colors.")
-
-            return colors
+            atexit.register(self._print_fps)
 
-        import_libs()
-        frame = ImageFunctions.convert(frame, format="OpenCV")
-
-        if self._frame_scaler is None:
-            _, width = frame.shape[0:2]
-            self._frame_scaler = width / self._image_processing_width
+    @property
+    def balls(self):
+        return {c: self.detectors[c].ball for c in self.detectors}
 
-        for c in parse_colors(color):
-            self.balls[c] = self.__find_most_likely_ball(
-                ball=self.balls.get(c), frame=frame, color=c
-            )
+    def add_color_hsv(self, color: str, lower: tuple, upper: tuple):
+        HSVColorRanges.add(color, lower, upper)
+        self.detectors[color] = SingleBallDetector(color, self.image_processing_width)
 
-        robot_view = frame.copy()
-        ball_data = DotDict({})
-        for ball_color, ball_object in self.balls.items():
-            ball_data[ball_color] = ball_object
-            self.__draw_ball_contrail(robot_view, ball_object)
-            if ball_object.found:
-                self.__draw_ball_position(robot_view, ball_object)
+    def _print_fps(self):
+        self._fps.stop()
 
-        ball_data["robot_view"] = ImageFunctions.convert(robot_view, self.format)
+    def __call__(
+        self,
+        frame,
+        color: Optional[Union[str, list]] = None,
+        hsv_limits: Optional[tuple] = None,
+    ):
+        if color and hsv_limits:
+            raise ValueError("Cannot specify both color and hsv_limits")
+
+        if hsv_limits:
+            assert len(hsv_limits) == 2, "hsv_limits must be a tuple of 2 elements"
+            hsv_lower, hsv_upper = [np.array(value) for value in hsv_limits]
+            # check if the color already exists
+            color_name = HSVColorRanges.get_color_for_hsv_limits(hsv_lower, hsv_upper)
+            if color_name is None:
+                # if not, add it with a random name
+                color_name = f"pt_{np.random.randint(0, 1000)}"
+                self.add_hsv_color(color_name, hsv_lower, hsv_upper)
```

Comment:
This is missing a test case (it generates a custom color for limits not already in the HSVColorRanges class)